File: CPP/ByTier/F/2019_F_Max_Plus_Min_Plus_Size.cpp
// Problem: CF 2019 F - Max Plus Min Plus Size
// https://codeforces.com/contest/2019/problem/F

/*
F. Max Plus Min Plus Size
time limit per test
2 seconds
memory limit per test
256 megabytes
input
standard input
output
standard output

The problem involves selecting a subset of elements in an array such that no two selected elements are adjacent,
and the score is defined as: max(selected) + min(selected) + count(selected).

This solution uses a greedy approach with a disjoint-set (DSU) data structure.
We process elements in decreasing order of value, merging intervals and tracking max/min information.

Time Complexity: O(n log n) due to sorting and DSU operations (amortized).
Space Complexity: O(n) for the DSU, arrays, and auxiliary structures.

Algorithms/Techniques: 
- Disjoint Set Union (DSU)
- Greedy algorithm
- Sorting
*/

#include <algorithm>
#include <ctype.h>
#include <stdio.h>
#include <string.h>
#define mem(_B, _E, _S) (memset((_B), (_E), (_S) * sizeof(*_B)))
char Getchar() {
  static char buf[1 << 22], *p1 = buf, *p2 = buf;
  return p1 == p2 && (p2 = (p1 = buf) + fread(buf, 1, 1 << 22, stdin), p1 == p2)
             ? EOF
             : *p1++;
}
void read(int &t) {
  t = 0;
  char ch = Getchar();
  while (!isdigit(ch))
    ch = Getchar();
  while (isdigit(ch))
    t = t * 10 + ch - '0', ch = Getchar();
}
const int N = 2e5 + 10;
using LL = long long;
int a[N], id[N];
int res, tag; // res stores contribution from size, tag tracks max value of red elements
int fmx[2][N], siz[N], dsu[N], vis[N];
int mx; // maximum element in the current array
int Find(int x) { return x == dsu[x] ? x : dsu[x] = Find(dsu[x]); }
int chkmx(int x) { return (siz[x] & 1) ? fmx[1][x] : (fmx[0][x] | fmx[1][x]); }
void Upd(int x, int t) {
  res += t * (siz[x] + 1) / 2; // Update contribution from size
  tag += t * chkmx(x);         // Update max element in the group
}
void Merge(int x, int y) {
  x = Find(x), y = Find(y);
  if (x == y)
    return;
  Upd(x, -1), Upd(y, -1); // Remove contributions of old components before merging
  int o = siz[x] & 1;
  fmx[0][x] |= fmx[o][y];   // Update max in even positions
  fmx[1][x] |= fmx[o ^ 1][y]; // Update max in odd positions accordingly
  siz[x] += siz[y];        // Merge sizes
  dsu[y] = x;              // Union operation
  Upd(x, 1);               // Add updated contribution of merged component
}
void solve() {
  int n;
  read(n);
  for (int i = 1; i <= n; ++i)
    read(a[i]);
  for (int i = 1; i <= n; ++i)
    id[i] = i, dsu[i] = i;
  // Sort indices by their corresponding values in descending order
  std::sort(id + 1, id + n + 1, [&](int i, int j) { return a[i] > a[j]; });
  mx = a[id[1]], res = tag = 0; // mx is highest value, initialize results
  int ans = 0;
  mem(fmx[0], 0, n + 1), mem(fmx[1], 0, n + 1); // Reset the max tracking arrays
  mem(vis, 0, n + 1);
  std::fill(siz + 1, siz + n + 1, 1); // Initially, all sets have size 1
  for (int i = 1; i <= n; ++i) {
    int x = id[i];
    vis[x] = 1;
    fmx[1][x] = a[x] == mx; // Mark if the element is max
    Upd(x, 1);             // Add contribution of single-node component
    if (x > 1 && vis[x - 1])
      Merge(x - 1, x);
    if (x < n && vis[x + 1])
      Merge(x, x + 1);
    ans = std::max(ans, mx + a[x] + res - (tag == 0)); // Update answer with current best
  }
  printf("%d\n", ans);
}
int main() {
  int T;
  read(T);
  while (T--)
    solve();
}


// https://github.com/VaHiX/codeForces/

File: Python/ByRound/2019/2019_F_Max_Plus_Min_Plus_Size.py
# Problem: CF 2019 F - Max Plus Min Plus Size
# https://codeforces.com/contest/2019/problem/F

"""
F. Max Plus Min Plus Size

Purpose:
This code solves a dynamic programming problem where we need to select non-adjacent elements from an array such that 
the sum of (max element, min element, and count) of selected elements is maximized.

Approach:
The solution uses a Union-Find (Disjoint Set Union) data structure combined with a greedy strategy. 
We process elements in decreasing order of value and merge adjacent valid segments to maintain connectivity,
tracking how many "even" segments exist (where even_max or odd_max properties are active).
This avoids recomputing from scratch for each new selection.

Algorithms:
- Union-Find (Disjoint Set Union) with path compression
- Sorting elements in decreasing order
- Greedy merging of adjacent valid segments

Time Complexity: O(n log n)
Space Complexity: O(n)

"""

import sys
from operator import itemgetter

input = sys.stdin.readline
rn = lambda: int(input())
rs = lambda: input().strip()
rl = lambda: list(map(int, input().split()))


def solve():
    n, a = rn(), rl()
    # Sort elements by value in descending order along with their indices
    b = sorted([(a[i], i) for i in range(n)], key=itemgetter(0), reverse=True)
    max_val = b[0][0]  # Maximum value in the array
    fa, node = list(range(n)), [None] * n  # Union-Find parent array and node info storage
    get_size = lambda node: node[3] - node[2] + 1  # Calculate size of a segment (node info)
    contain_even_count, current_cnt, res = 0, 0, 0  # Counters for tracking results

    def contain_even(node):
        # Check if a node contains 'even' segments based on max even/odd flags
        even_max, odd_max, l, r, _ = node
        return (even_max and ~(l & r) & 1) or (odd_max and (l | r) & 1)

    def find(x):
        # Find root with path compression
        if fa[x] != x:
            fa[x] = find(fa[x])
            return fa[x]
        else:
            return x

    def merge(x, y):
        nonlocal contain_even_count, current_cnt
        fx, fy = find(x), find(y)
        if fx == fy:
            return
        # Ensure fx is the smaller group to optimize merging
        if node[fx][4] > node[fy][4]:
            fx, fy = fy, fx
        fa[fx] = fy  # Merge smaller group into larger one
        node_fx, node_fy = node[fx], node[fy]
        even_max_fx, odd_max_fx, l_fx, r_fx, res_fx = node_fx
        even_max_fy, odd_max_fy, l_fy, r_fy, res_fy = node_fy
        # Update merged node properties
        node[fy] = (
            even_max_fx or even_max_fy,
            odd_max_fx or odd_max_fy,
            min(l_fx, l_fy),
            max(r_fx, r_fy),
            (max(r_fx, r_fy) - min(l_fx, l_fy) + 2) >> 1,
        )
        # Update even count and total current count
        contain_even_count += (
            contain_even(node[fy]) - contain_even(node_fx) - contain_even(node_fy)
        )
        current_cnt += node[fy][4] - res_fx - res_fy

    for v, i in b:
        if v == max_val:
            # Initialize node for maximum value with special even property
            node[i] = (~i & 1, i & 1, i, i, 1)
            contain_even_count += 1
        else:
            # Initialize node for other values
            node[i] = (False, False, i, i, 1)
        current_cnt += 1  # Count the inclusion of this element
        # Try to merge with left neighbor if valid
        if i > 0 and a[i - 1] >= v and node[i - 1]:
            merge(i - 1, i)
        # Try to merge with right neighbor if valid
        if i < n - 1 and a[i + 1] >= v and node[i + 1]:
            merge(i, i + 1)
        # Update global result
        res = max(res, current_cnt + v + max_val - (contain_even_count == 0))
    print(res)


[solve() for _ in range(rn())]


# https://github.com/VaHiX/codeForces/